### PR TITLE
Schema: change sort order of history event type enum

### LINKED ISSUE #614: Prepare `history` table for order by `event_time desc, event_type desc`
This includes:

* Update `history.event_type` to reflect the correlation of different events (e.g. `downtime_start -> downtime_end`)
* Add an index for `history.event_time desc, history.event_type desc` (not sure if the direction matters being in the definition)

### PR DESCRIPTION
**Pull request by julianbrost**
This improves the resulting sort order when `ORDER BY event_time, event_type` is used. `state_change` comes first as it can cause many of the other events like trigger downtimes, remove acknowledgements and send notifications. Similarly, `notification` comes last as any other event can result in a notification. This will result in history events for scenarios like state changes, triggers downtime, sends downtime start notification being sorted in that order.

Apart from that, end events sort before the corresponding start events as any ack/comment/downtime/flapping period should last for more than a millisecond, therefore if there should be two events within the same millisecond, the end event corresponds to the older period and is sorted first. The remaining order is by impact and cause, see inline comment or https://github.com/Icinga/icingadb/pull/626#discussion_r1284399172.

Both schema upgrades will lock the `history` table for a few minutes, blocking other read and write access including by the Icinga DB daemon and Icinga DB Web. On my (totally representative) laptop, this was about 2 minutes for 19M rows in MariaDB and 8 minutes for 22M rows in PostgreSQL (worst-case times, I've also seen it faster on both).

If you're wondering about why the default value for PostgreSQL is changed: I don't think that this does anything useful, it's just the first enum value to emulate that default behavior from MySQL.

### For the review
* It's better if more people look at this, we really don't want to have to change this again.
* Please double check the ordering, ideally make your own thoughts and check if they are compatible with my suggestion.

fixes #614
refs https://github.com/Icinga/icingadb-web/issues/587

### CODE REVIEW

File: schema/mysql/upgrades/1.2.0.sql [hunk 1]
```diff
@@ -14,3 +14,16 @@ ALTER TABLE servicegroup
...
+
+-- The following sequence of statements changes the type of history.event_type like the following statement would:
+--
+--   ALTER TABLE history MODIFY event_type enum('state_change', 'ack_clear', 'comment_remove', 'downtime_end', 'flapping_end', 'ack_set', 'comment_add', 'downtime_start', 'flapping_start', 'notification') NOT NULL;
+--
+-- It's just much faster to add a second column, copy the column using an UPDATE statement and then replace the
+-- old column with the one just generated. Table locking ensures that no other connection inserts data in the meantime.
+LOCK TABLES history WRITE;
+ALTER TABLE history ADD COLUMN event_type_new enum('state_change', 'ack_clear', 'comment_remove', 'downtime_end', 'flapping_end', 'ack_set', 'comment_add', 'downtime_start', 'flapping_start', 'notification') NOT NULL AFTER event_type;
+UPDATE history SET event_type_new = event_type;
```
Comment (Al2Klimov) [inline schema/mysql/upgrades/1.2.0.sql:26]: Seems to work w/o explicit string casts:

```
MariaDB [idb]> create table history ( event_type enum('notification','state_change','downtime_start', 'downtime_end','comment_add','comment_remove','flapping_start','flapping_end','ack_set','ack_clear') NOT NULL );
Query OK, 0 rows affected (0,014 sec)

MariaDB [idb]> insert into history values ();
Query OK, 1 row affected (0,008 sec)

MariaDB [idb]> select * from history;
+--------------+
| event_type   |
+--------------+
| notification |
+--------------+
1 row in set (0,003 sec)

MariaDB [idb]> LOCK TABLES history WRITE;
Query OK, 0 rows affected (0,003 sec)

MariaDB [idb]> ALTER TABLE history ADD COLUMN event_type_new enum('state_change', 'ack_clear', 'comment_remove', 'downtime_end', 'flapping_end', 'ack_set', 'comment_add', 'downtime_start', 'flapping_start', 'notification') NOT NULL AFTER event_type;
Query OK, 0 rows affected (0,017 sec)
Records: 0  Duplicates: 0  Warnings: 0

MariaDB [idb]> UPDATE history SET event_type_new = event_type;
Query OK, 1 row affected (0,008 sec)
Rows matched: 1  Changed: 1  Warnings: 0

MariaDB [idb]> ALTER TABLE history DROP COLUMN event_type,
    ->     CHANGE COLUMN event_type_new event_type enum('state_change', 'ack_clear', 'comment_remove', 'downtime_end', 'flapping_end', 'ack_set', 'comment_add', 'downtime_start', 'flapping_start', 'notification') NOT NULL;
Query OK, 0 rows affected (0,025 sec)
Records: 0  Duplicates: 0  Warnings: 0

MariaDB [idb]> UNLOCK TABLES;
Query OK, 0 rows affected (0,003 sec)

MariaDB [idb]> select * from history;
+--------------+
| event_type   |
+--------------+
| notification |
+--------------+
1 row in set (0,004 sec)

MariaDB [idb]>
```

I can't complain.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
For the record: the change as such is not super critical. (We're talking about two events for the same object in the same millisecond.) Neither is the schema upgrade which is just for consistency.

# Suggestion

Defer this to 1.2.0 to make the bugfix release more attractive.

**Comment by julianbrost** [comment]
> (We're talking about two events for the same object in the same millisecond.)

Which, however isn't some esoteric edge case but something that actually happens as flexible downtime starts and notifications are written with the same timestamp as the corresponding state change. So pretty much the main goal of this change is to get this sorted properly in the history.

File: schema/pgsql/upgrades/1.2.0.sql [hunk 2]
```diff
@@ -14,3 +14,11 @@ CREATE INDEX idx_servicegroup_name_ci ON servicegroup(name_ci);
 COMMENT ON INDEX idx_servicegroup_display_name IS 'Servicegroup list filtered/ordered by display_name';
 COMMENT ON INDEX idx_servicegroup_name_ci IS 'Servicegroup list filtered using quick search';
 COMMENT ON INDEX idx_servicegroup_name IS 'Host/service/service group list filtered by service group name; Servicegroup detail filter';
+
+ALTER TYPE history_type RENAME TO history_type_old;
+CREATE TYPE history_type AS ENUM ( 'state_change', 'ack_clear', 'comment_remove', 'downtime_end', 'flapping_end', 'ack_set', 'comment_add', 'downtime_start', 'flapping_start', 'notification' );
+ALTER TABLE history
+  ALTER COLUMN event_type DROP DEFAULT,
+  ALTER COLUMN event_type TYPE history_type USING event_type::text::history_type,
+  ALTER COLUMN event_type SET DEFAULT 'state_change'::history_type;
```
[the remark this answers is not in the document]
Comment (julianbrost) [reply schema/pgsql/upgrades/1.2.0.sql:23]: `UPDATE history SET event_type_new = event_type::text::history_type;` is already running for 15 minutes on the same database and not finished yet, so I don't think this will help.

[the remark this answers is not in the document]
Comment (julianbrost) [reply schema/pgsql/upgrades/1.2.0.sql:23]: Finished after 70 minutes, so yeah, not the way to go.

File: schema/mysql/schema.sql [hunk 3]
```diff
@@ -1269,7 +1269,7 @@ CREATE TABLE history (
   flapping_history_id binary(20) DEFAULT NULL COMMENT 'flapping_history.id',
   acknowledgement_history_id binary(20) DEFAULT NULL COMMENT 'acknowledgement_history.id',
 
-  event_type enum('notification','state_change','downtime_start', 'downtime_end','comment_add','comment_remove','flapping_start','flapping_end','ack_set','ack_clear') NOT NULL,
+  event_type enum('state_change', 'ack_clear', 'comment_remove', 'downtime_end', 'flapping_end', 'ack_set', 'comment_add', 'downtime_start', 'flapping_start', 'notification') NOT NULL,
```
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply schema/mysql/schema.sql:1277]: I meant a COMMENT, but this is acceptable. Now as I think about it: we use COMMENTs to answer the _what_, not the _why_ anyway.

**Review by Al2Klimov — APPROVED** [review]

**Review by nilmerg — APPROVED** [review]
Can only comment on the ordering, which is fine.

File: schema/mysql/schema.sql [hunk 4]
```diff
@@ -1269,7 +1269,12 @@ CREATE TABLE history (
   flapping_history_id binary(20) DEFAULT NULL COMMENT 'flapping_history.id',
   acknowledgement_history_id binary(20) DEFAULT NULL COMMENT 'acknowledgement_history.id',
 
-  event_type enum('notification','state_change','downtime_start', 'downtime_end','comment_add','comment_remove','flapping_start','flapping_end','ack_set','ack_clear') NOT NULL,
+  -- The enum values are ordered in a way that event_type provides a meaningful sort order for history entries with
+  -- the same event_time. state_change comes first as it can cause many of the other events like trigger downtimes,
+  -- remove acknowledgements and send notifications. Similarly, notification comes last as any other event can result
+  -- in a notification. End events sort before the corresponding start events as any ack/comment/downtime/flapping
+  -- period should last for more than a millisecond, therefore, the old period ends first and then the new one starts.
+  event_type enum('state_change', 'ack_clear', 'comment_remove', 'downtime_end', 'flapping_end', 'ack_set', 'comment_add', 'downtime_start', 'flapping_start', 'notification') NOT NULL,
```
Comment (lippserd) [inline schema/mysql/schema.sql:1277]: I would order as follows: `state_change`, `ack_clear`, `downtime_end`, `flapping_end`, `comment_remove`, `comment_add`, `flapping_start`, `downtime_start`, `ack_set`, `notification`.

Reasoning:

* Comments are informative.
* Flapping is automatic and changes mechanics.
* Downtimes are semi-automatic, require user action (or configuration) and change mechanics.
* Acks are pure user actions and change mechanics.

In short, the ordering indicates whether the event changes Icinga mechanics and whether its origin is Icinga or the user.

**Review by lippserd — CHANGES_REQUESTED** [review]

File: schema/mysql/schema.sql (hunk 4 above)
Comment (julianbrost) [reply schema/mysql/schema.sql:1277]: Only changes the ordering within the start and end events (without mixing the two groups), so basically only changes where I've arbitrarily chosen to sort alphabetically, so no complaints from me.

> Reasoning: [...] In short, the ordering indicates whether the event changes Icinga mechanics and whether its origin is Icinga or the user.

So basically assigns the events some kind of importance (ack being the most important and comment the least), so that an `ack_set` will show up as the most recent entry (besides notifications).

**Review by lippserd — APPROVED** [review]

**Review by Al2Klimov — APPROVED** [review]
IF we can make the order EVEN smarter, I don't know how 👍
